### DISCUSSION TITLE
C++ 17 compilation error due to  return m_value.array->emplace_back peace of code

### DISCUSSION THREAD

**Discussion by pixout** [nlohmann/json#2438] [Q&A] [answered]
Hi,

I'm trying to compile json library with CMake flag set(CMAKE_CXX_STANDARD 17) and getting the following error.

``
#ifdef JSON_HAS_CPP_17
        return m_value.array->emplace_back(std::forward<Args>(args)...);
#else

nlohmann/json.hpp(5431): error C2440: 'return': cannot convert from 'void' to 'nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer,std::vector<unsigned char,std::allocator<_Ty>>> &' [Analytics.vcxproj]
          with
          [
              _Ty=unsigned char
          ]
``

Please possible somebody knows how to fix it

**Comment by pixout** [comment]
This happens when I'm using v140 build tools and the latest 2019 Visual studio

**Comment by nlohmann** [comment]
Can you please confirm that the error is actually in line

```cpp
return m_value.array->emplace_back(std::forward(args)...);
```

Looking at https://en.cppreference.com/w/cpp/container/deque/emplace_back, the return value is `reference` in C++17. So if the error occurs in C++17 mode on that line, then the STL of MSVC seems to be broken.

**Comment by pixout** [comment, marked as answer]
Yes, the error was on this line, but the root cause was in a different case. We use v140 toolset (from MSVC 2015 without C++17) for 2019 MSVC but in root CMake file specified CMAKE_CXX_STANDARD 17 and due to nlohman lib macros_scope.hpp condition the following:

```
// C++ language standard detection
#if (defined(__cplusplus) && __cplusplus >= 202002L) || (defined(_MSVC_LANG) && _MSVC_LANG >= 202002L)
    #define JSON_HAS_CPP_20
    #define JSON_HAS_CPP_17
    #define JSON_HAS_CPP_14
#elif (defined(__cplusplus) && __cplusplus >= 201703L) || (defined(_HAS_CXX17) && _HAS_CXX17 == 1) // fix for issue #464
    #define JSON_HAS_CPP_17
    #define JSON_HAS_CPP_14
#elif (defined(__cplusplus) && __cplusplus >= 201402L) || (defined(_HAS_CXX14) && _HAS_CXX14 == 1)
    #define JSON_HAS_CPP_14
#endif

```

Library set JSON_HAS_CPP_17 define but actually, there is no C++17 support

**Reply by nlohmann** [reply]
What do you propose to fix this?

**Reply by pixout** [reply]
We have made few following changes in the root CMake and now everything working like a charm.

```
	if(CMAKE_VS_PLATFORM_TOOLSET)
		if(CMAKE_VS_PLATFORM_TOOLSET STRGREATER "v140" )
			set(CMAKE_CXX_STANDARD 17)
		endif()
	endif()

```

**Reply by nlohmann** [reply]
Is there anything I can change in the library code? I am not using MSVC, so I don't really understand all this. PRs welcome ;-)

**Comment by nlohmann** [comment]
Do you need further assistance with this issue?

**Reply by pixout** [reply]
No, Thanks, issue could be close
